Added memory_info table for Linux

nerddotcat opened the pull request:
Added a memory_info table for Linux machines based on [this issue](https://github.com/facebook/osquery/issues/2028).  The `/proc/meminfo` file is read and a single row containing memory information is stored in the table.

I kept the memory in units of kilobytes just like the `meminfo` file uses.
I was wondering if it's fine to keep the data type as INTEGER or if I should use BIGINT. It looks like the INTEGER type is 64 bits for sqlite so I don't think this is an issue.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

PoppySeedPlehzr commented:
ok to test

osqueryer commented:
Build origin/pr/2282/merge (Job: 2924) is good! :+1:

theopolis commented on specs/linux/memory_info.table:
```diff
@@ -0,0 +1,17 @@
+table_name("memory_info")
+
+description("Main memory information, in kilobytes)")
```
nitpick, there's a trailing ")" here.

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,81 @@
+      boost::split(
+          tokens, line, boost::is_any_of("\t "), boost::token_compress_on);
+      if (line.find("MemTotal:") == 0) {
+        r["memory_total"] = INTEGER(std::stoi(tokens[1]));
```
For each of these, can you `std::stoi(tokens[1]) * 1024` so the reporting can be in bytes?

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,81 @@
+      std::vector<std::string> tokens;
+      boost::split(
+          tokens, line, boost::is_any_of("\t "), boost::token_compress_on);
+      if (line.find("MemTotal:") == 0) {
```
Instead of writing a large set of conditionals here, you can:
```cpp
std::map<std::string, std::string> indexes = {
  {"memory_total", "MemTotal"},
  {"memory_free", "MemFree"},
};
```

Then:
```cpp
for (const auto& p : indexes) {
  if (line.find(p.second) == 0) {
    r[p.first] = ...;
    break;
  }
}
```

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,81 @@
+QueryData getMemoryInfo(QueryContext& context) {
+  QueryData results;
+  Row r;
+  r["memory_total"] = INTEGER(-1);
```
For values that aren't present, we should bias towards reporting an SQLite `NULL`. Right now that's just "not filling in data" ;)

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,81 @@
+
+QueryData getMemoryInfo(QueryContext& context) {
+  QueryData results;
+  Row r;
```
Place this `Row r` in your loop. You will construct the map each time, and assure that there's no "stale" column from a previous iteration.

osqueryer commented:
Build origin/pr/2282/merge (Job: 2942) is good! :+1:

osqueryer commented:
Build origin/pr/2282/merge (Job: 2943) is good! :+1:

osqueryer commented:
Build origin/pr/2282/merge (Job: 2944) is good! :+1:

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,62 @@
+namespace osquery {
+namespace tables {
+
+const std::string memInfoPath = {"/proc/meminfo"};
```
Sorry I didn't catch/note this before, but for global `const` we begin each variable with `k` so this would become: `kMemInfoPath`.

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,62 @@
+
+const std::string memInfoPath = {"/proc/meminfo"};
+
+std::map<std::string, std::string> meminfoMap = {
```
This can be `const` too, and see above: `kMemInfoMap`.

theopolis commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,62 @@
+const std::string memInfoPath = {"/proc/meminfo"};
+
+std::map<std::string, std::string> meminfoMap = {
+    {"memory_total", "MemTotal:"},
```
I don't *think* you need to include the trailing `:` here for the matching, right?

zwass commented on osquery/tables/system/linux/memory_info.cpp:
```diff
@@ -0,0 +1,62 @@
+      // Look for mapping
+      for (const auto& singleMap : meminfoMap) {
+        if (line.find(singleMap.second) == 0) {
+          r[singleMap.first] = INTEGER(std::stol(tokens[1]) * 1024l);
```
@theopolis: Outside the context of this PR, maybe we should add a set of constants somewhere for expressing things like `bytes_in_kb`.

osqueryer commented:
Build origin/pr/2282/merge (Job: 2947) is good! :+1:

nerddotcat commented:
Here's what I updated yesterday:
- Changed memory units from kilobytes to bytes
- Created map for columns to meminfo entries to remove large amount of if statements
- Columns now default to null instead of -1
- Fixed a comment

Today I fixed constant variable names.

> Place this Row r in your loop. You will construct the map each time, and assure that there's no "stale" column from a previous iteration.

I don't understand what you're getting at here. Since this only keeps track of a single file per machine, there will only be one row unless you're aggregating from multiple machines. The loop goes over each line in the meminfo, so if you moved creating the row inside the loop, you would get a bunch of rows but it would only have information for one column per row.

> I don't think you need to include the trailing : here for the matching, right?

The trailing `:` isn't necessary for the values we are currently grabbing, but it could be a problem.
Here's part of the `meminfo` file from my machine:
`Active:          1095412 kB`
`Inactive:         499180 kB`
`Active(anon):     584992 kB`
`Inactive(anon):   339784 kB`


`Active` will also match with `Active(anon)` and I thought it would be safer to include the ending `:`

theopolis commented:
Solid work, @PoppySeedPlehzr and I just spoke about the `\t` split and skipped check for the size of the resultant vector and we decided it's no big problem. Since this is a pseudo-file, not having the correct format would mean a change in the proc API. The effect for osquery would be a failure line noting the exception, which is valuable information over checking size and emitting a verbose line when the size is not `>= 2`.
